### PR TITLE
Hashtable

### PR DESCRIPTION
**Pull request by wildwoodwaltz**

### CODE REVIEW

**Comment by wildwoodwaltz** [comment]
Testing 
- [x]    Setting a key/value to your hashtable results in the value being in the data structure
- [x]      Retrieving based on a key returns the value stored
- [x]     Successfully returns null for a key that does not exist in the hashtable
- [x]      Successfully returns a list of all unique keys that exist in the hashtable
- [x]      Successfully handle a collision within the hashtable
- [x]      Successfully retrieve a value from a bucket within the hashtable that has a collision
- [x]      Successfully hash a key to an in-range value

File: python/docs/hashtable/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+Create a Hash Table Data Structure with set, get, keys, contains, hash.
+
+## Approach & Efficiency
+
+Built using TDD to verify as it was built well from the ground up.
+
+Efficiency:
+
+Generally this procedure will take an O(1) Time and space complexity if everything goes well, however... Due to the instantiation of a linked list to handle collisions.
```
Comment (chloenott) [inline python/docs/hashtable/README.md:13]: Nice! Good detail to point out here of O(1) vs O(n) with worst case collisions.

File: python/data_structures/hashtable.py [hunk 2]
```diff
@@ -1,12 +1,94 @@
...
+
+    def keys(self):
+        keys = set()
+        for bucket in self._buckets:
+            try:
+                if bucket:
+                    container = bucket.head
+                while bucket:
+                    keys.add(container.value[0])
+                    print(keys)
```
Comment (chloenott) [inline python/data_structures/hashtable.py:63]: Can probably delete this print now that building/debugging is over.

File: python/data_structures/hashtable.py [hunk 3]
```diff
@@ -1,12 +1,94 @@
...
     """
     Put docstring here
     """
+    def __init__(self, size=1024):
+        self.size = size
+        self._buckets = [None] * size
+
+    def set(self, key, value):
+        '''
+        Sets item into hashtable by first calculating hash, then if it doesn't have anything in it, then instantiate a linked list to hold valuse in case of possible collision.
```
Comment (chloenott) [inline python/data_structures/hashtable.py:14]: I like the docstrings! Recommend adding them to the methods below missing them as well. Could probably copy/paste the readme's comments for each method over.

File: python/data_structures/hashtable.py [hunk 4]
```diff
@@ -1,12 +1,94 @@
...
+        
+        '''
+        Checks to see if key exists and if does, updates that key, value pair
+            '''
+        # current = self._buckets[idx].head
+
+        # while current:
+        #     if current.value[0] == key:
+        #         current.value = (key, value)
+        #     current = current.next
```
Comment (chloenott) [inline python/data_structures/hashtable.py:31]: Recommend deleting unused comments for code cleanup.

**Review by chloenott — COMMENTED** [review]
All looks good! Left a few comments on minor things to consider for next time.

**Pull request merged**